File: ios/LokmaCore/Sources/LokmaCore/DepthVolumeEngine.swift
// DepthVolumeEngine — Tier 2 LiDAR depth-integrated volume above a fitted plane.
//
// The Swift mirror of `lokma/geometry/depth_volume_engine.py`. Pure, deterministic,
// and arithmetic-identical (same ring band, same explicit 3×3 Cramer solve, same
// median/MAD trim, same integration order) so the golden-vector parity holds.
//
// Math: fit the support plane z = a·x + b·y + c from the depth ring around the
// food, then
//     V_mm³ = (1/(fx·fy)) · Σ_food max(0, z_plane(xᵢ,yᵢ) − zᵢ) · zᵢ²
// which reduces to height × footprint-area for a flat top-down plate and stays
// correct under tilt. Operates in depth-map pixel space (never upsamples depth).
import Foundation

public struct DepthVolumeParams: Sendable {
    public var minCoverage: Double = 0.5
    public var bboxMarginFrac: Double = 0.25
    public var madK: Double = 3.0
    public var minRingPoints: Int = 30
    public var heightClampMm: Double = 0.0
    public init() {}
}

public struct DepthVolumeResult: Sendable, Equatable {
    public let volumeCm3: Double
    public let coverage: Double
    public let planeResidualMm: Double
    public let method: String

    public init(volumeCm3: Double, coverage: Double, planeResidualMm: Double, method: String = "depth_plane") {
        self.volumeCm3 = volumeCm3
        self.coverage = coverage
        self.planeResidualMm = planeResidualMm
        self.method = method
    }
}

public struct DepthVolumeEngine: Sendable {
    public init() {}

    private static let detEps = 1e-9
    private static let madEps = 1e-9

    /// Integrate food volume (cm³) above the fitted support plane, or nil when the
    /// depth is too sparse / the ring too small / the plane singular (caller falls
    /// back to the scalar volumetric path). `depthMm`/`mask` are row-major, length
    /// `width*height`; depth ≤ 0 (or non-finite) marks an invalid pixel.
    public func integrate(
        depthMm: [Double], mask: [Double], width: Int, height: Int,
        fx: Double, fy: Double, cx: Double, cy: Double,
        params: DepthVolumeParams = DepthVolumeParams()
    ) -> DepthVolumeResult? {
        let n = width * height
        guard n > 0, depthMm.count == n, mask.count == n, fx != 0, fy != 0 else { return nil }

        // Food bounding box.
        var foodTotal = 0
        var minU = width, maxU = -1, minV = height, maxV = -1
        for v in 0..<height {
            for u in 0..<width where mask[v * width + u] >= 0.5 {
                foodTotal += 1
                if u < minU { minU = u }
                if u > maxU { maxU = u }
                if v < minV { minV = v }
                if v > maxV { maxV = v }
            }
        }
        guard foodTotal > 0 else { return nil }

        let mx = Int(params.bboxMarginFrac * Double(maxU - minU + 1) + 0.5)
        let my = Int(params.bboxMarginFrac * Double(maxV - minV + 1) + 0.5)
        let eMinU = max(0, minU - mx), eMaxU = min(width - 1, maxU + mx)
        let eMinV = max(0, minV - my), eMaxV = min(height - 1, maxV + my)

        // Ring = valid, non-food pixels inside the expanded bbox (row-major order).
        var rx: [Double] = [], ry: [Double] = [], rz: [Double] = []
        for v in eMinV...eMaxV {
            for u in eMinU...eMaxU {
                let i = v * width + u
                let z = depthMm[i]
                guard z.isFinite, z > 0, mask[i] < 0.5 else { continue }
                rx.append((Double(u) - cx) / fx * z)
                ry.append((Double(v) - cy) / fy * z)
                rz.append(z)
            }
        }
        guard rx.count >= params.minRingPoints else { return nil }

        guard var plane = Self.solvePlane(rx, ry, rz) else { return nil }

        // One MAD-trim of ring residuals, then refit.
        var resid = [Double](repeating: 0, count: rx.count)
        for j in 0..<rx.count { resid[j] = plane.a * rx[j] + plane.b * ry[j] + plane.c - rz[j] }
        let med = Self.median(resid)
        var absdev = [Double](repeating: 0, count: resid.count)
        for j in 0..<resid.count { absdev[j] = abs(resid[j] - med) }
        let mad = Self.median(absdev)
        if mad > Self.madEps {
            let cutoff = params.madK * mad
            var kx: [Double] = [], ky: [Double] = [], kz: [Double] = []
            for j in 0..<rx.count where absdev[j] <= cutoff {
                kx.append(rx[j]); ky.append(ry[j]); kz.append(rz[j])
            }
            if kx.count >= 3, let refit = Self.solvePlane(kx, ky, kz) {
                plane = refit; rx = kx; ry = ky; rz = kz
            }
        }

        var sse = 0.0
        for j in 0..<rx.count {
            let r = plane.a * rx[j] + plane.b * ry[j] + plane.c - rz[j]
            sse += r * r
        }
        let planeResidual = rx.isEmpty ? 0.0 : (sse / Double(rx.count)).squareRoot()

        // Coverage gate.
        var foodValid = 0
        for i in 0..<n where mask[i] >= 0.5 && depthMm[i].isFinite && depthMm[i] > 0 { foodValid += 1 }
        let coverage = Double(foodValid) / Double(foodTotal)
        guard coverage >= params.minCoverage else { return nil }

        // Integrate V = (1/(fx·fy)) Σ max(0, z_plane − z) · z².
        var sumTerm = 0.0
        for v in 0..<height {
            for u in 0..<width {
                let i = v * width + u
                guard mask[i] >= 0.5 else { continue }
                let z = depthMm[i]
                guard z.isFinite, z > 0 else { continue }
                let x = (Double(u) - cx) / fx * z
                let y = (Double(v) - cy) / fy * z
                let h = (plane.a * x + plane.b * y + plane.c) - z
                if h > params.heightClampMm { sumTerm += h * z * z }
            }
        }
        let volumeCm3 = (sumTerm / (fx * fy)) / 1000.0
        return DepthVolumeResult(volumeCm3: volumeCm3, coverage: coverage, planeResidualMm: planeResidual)
    }

    /// Convenience over a `DepthSample` (mirrors `integrate_sample` in Python).
    public func integrate(_ sample: DepthSample, params: DepthVolumeParams = DepthVolumeParams()) -> DepthVolumeResult? {
        integrate(depthMm: sample.depthMm, mask: sample.mask, width: sample.width, height: sample.height,
                  fx: sample.fx, fy: sample.fy, cx: sample.cx, cy: sample.cy, params: params)
    }

    // MARK: - deterministic helpers (mirror models in depth_volume_engine.py)

    static func median(_ xs: [Double]) -> Double {
        if xs.isEmpty { return 0.0 }
        let s = xs.sorted()
        let n = s.count, mid = n / 2
        return n % 2 == 1 ? s[mid] : 0.5 * (s[mid - 1] + s[mid])
    }

    private static func det3(_ m: [[Double]]) -> Double {
        m[0][0] * (m[1][1] * m[2][2] - m[1][2] * m[2][1])
            - m[0][1] * (m[1][0] * m[2][2] - m[1][2] * m[2][0])
            + m[0][2] * (m[1][0] * m[2][1] - m[1][1] * m[2][0])
    }

    /// Least-squares z = a·x + b·y + c via 3×3 normal equations + Cramer's rule.
    static func solvePlane(_ xs: [Double], _ ys: [Double], _ zs: [Double]) -> (a: Double, b: Double, c: Double)? {
        var sxx = 0.0, syy = 0.0, sxy = 0.0, sx = 0.0, sy = 0.0, sxz = 0.0, syz = 0.0, sz = 0.0
        let n = xs.count
        for i in 0..<n {
            let x = xs[i], y = ys[i], z = zs[i]
            sxx += x * x; syy += y * y; sxy += x * y
            sx += x; sy += y; sxz += x * z; syz += y * z; sz += z
        }
        let A = [[sxx, sxy, sx], [sxy, syy, sy], [sx, sy, Double(n)]]
        let b = [sxz, syz, sz]
        let det = det3(A)
        if abs(det) < detEps { return nil }
        let aa = [[b[0], A[0][1], A[0][2]], [b[1], A[1][1], A[1][2]], [b[2], A[2][1], A[2][2]]]
        let ab = [[A[0][0], b[0], A[0][2]], [A[1][0], b[1], A[1][2]], [A[2][0], b[2], A[2][2]]]
        let ac = [[A[0][0], A[0][1], b[0]], [A[1][0], A[1][1], b[1]], [A[2][0], A[2][1], b[2]]]
        return (det3(aa) / det, det3(ab) / det, det3(ac) / det)
    }
}

File: lokma/geometry/depth_volume_engine.py
"""Tier 2: LiDAR depth-integrated food volume above a fitted support plane.

Pure-numpy and deterministic — the Swift `DepthVolumeEngine` mirrors this exact
arithmetic for golden-vector parity (no cv2/scipy). Operates in depth-map pixel
space; never upsamples depth.

Math (derived in the plan): with the support plane fit as ``z = a·x + b·y + c``
from the depth ring around the food, the perpendicular-height × plane-footprint
integral collapses (the √(a²+b²+1) foreshortening terms cancel) to

    V_mm³ = (1 / (fx·fy)) · Σ_food  max(0, z_plane(xᵢ,yᵢ) − zᵢ) · zᵢ²

which reduces to exactly height × footprint-area for a flat top-down plate and
stays correct under tilt. Back-projection uses ``x = (u−cx)/fx·z`` (camera mm,
perpendicular-Z convention).
"""

from __future__ import annotations

from dataclasses import dataclass

import numpy as np

_DET_EPS = 1e-9
_MAD_EPS = 1e-9


@dataclass(frozen=True)
class DepthVolumeParams:
    """Gating + robustness knobs (mirrored verbatim in Swift)."""

    min_coverage: float = 0.5       # fraction of food pixels needing valid depth
    bbox_margin_frac: float = 0.25  # ring band width as a fraction of the food bbox
    mad_k: float = 3.0              # MAD multiplier for ring outlier rejection
    min_ring_points: int = 30       # below this the plane fit isn't trustworthy
    height_clamp_mm: float = 0.0    # ignore surface points at/below this height
    max_food_height_mm: float = 250.0  # reject if the median food height exceeds this
    #                                    (no near support plane → V would explode)


@dataclass(frozen=True)
class DepthVolumeResult:
    volume_cm3: float
    coverage: float
    plane_residual_mm: float
    method: str = "depth_plane"


def _median(values: list[float]) -> float:
    s = sorted(values)
    n = len(s)
    if n == 0:
        return 0.0
    mid = n // 2
    if n % 2 == 1:
        return s[mid]
    return 0.5 * (s[mid - 1] + s[mid])


def _det3(m) -> float:
    return (
        m[0][0] * (m[1][1] * m[2][2] - m[1][2] * m[2][1])
        - m[0][1] * (m[1][0] * m[2][2] - m[1][2] * m[2][0])
        + m[0][2] * (m[1][0] * m[2][1] - m[1][1] * m[2][0])
    )


def _solve_plane(xs: list[float], ys: list[float], zs: list[float]):
    """Least-squares ``z = a·x + b·y + c`` via 3×3 normal equations + Cramer's rule.

    Sequential accumulation (not ``np.linalg.lstsq``) so Swift can reproduce the
    coefficients bit-for-bit. Returns ``(a, b, c)`` or ``None`` when singular.
    """
    sxx = syy = sxy = sx = sy = sxz = syz = sz = 0.0
    n = len(xs)
    for i in range(n):
        x = xs[i]
        y = ys[i]
        z = zs[i]
        sxx += x * x
        syy += y * y
        sxy += x * y
        sx += x
        sy += y
        sxz += x * z
        syz += y * z
        sz += z
    A = [[sxx, sxy, sx], [sxy, syy, sy], [sx, sy, float(n)]]
    b = [sxz, syz, sz]
    det = _det3(A)
    if abs(det) < _DET_EPS:
        return None
    aa = [[b[0], A[0][1], A[0][2]], [b[1], A[1][1], A[1][2]], [b[2], A[2][1], A[2][2]]]
    ab = [[A[0][0], b[0], A[0][2]], [A[1][0], b[1], A[1][2]], [A[2][0], b[2], A[2][2]]]
    ac = [[A[0][0], A[0][1], b[0]], [A[1][0], A[1][1], b[1]], [A[2][0], A[2][1], b[2]]]
    return (_det3(aa) / det, _det3(ab) / det, _det3(ac) / det)


class DepthVolumeEngineService:
    """Integrate food volume above a deterministically fitted support plane."""

    def integrate(
        self,
        depth_mm,
        mask,
        width: int,
        height: int,
        fx: float,
        fy: float,
        cx: float,
        cy: float,
        params: DepthVolumeParams | None = None,
    ) -> DepthVolumeResult | None:
        p = params or DepthVolumeParams()
        if width <= 0 or height <= 0 or fx == 0.0 or fy == 0.0:
            return None
        depth = np.asarray(depth_mm, dtype=np.float64).reshape(height, width)
        m = np.asarray(mask, dtype=np.float64).reshape(height, width)

        valid = np.isfinite(depth) & (depth > 0.0)
        food = m >= 0.5
        food_total = int(food.sum())
        if food_total == 0:
            return None

        # Food bounding box, expanded by a margin to form the ring band.
        ys, xs = np.nonzero(food)
        min_u, max_u = int(xs.min()), int(xs.max())
        min_v, max_v = int(ys.min()), int(ys.max())
        mx = int(p.bbox_margin_frac * (max_u - min_u + 1) + 0.5)
        my = int(p.bbox_margin_frac * (max_v - min_v + 1) + 0.5)
        e_min_u, e_max_u = max(0, min_u - mx), min(width - 1, max_u + mx)
        e_min_v, e_max_v = max(0, min_v - my), min(height - 1, max_v + my)

        # Ring = valid, non-food pixels inside the expanded bbox.
        ring = np.zeros((height, width), dtype=bool)
        ring[e_min_v:e_max_v + 1, e_min_u:e_max_u + 1] = True
        ring &= valid & (~food)
        rv, ru = np.nonzero(ring)
        if rv.size < p.min_ring_points:
            return None

        rz = depth[rv, ru]
        rx = (ru.astype(np.float64) - cx) / fx * rz
        ry = (rv.astype(np.float64) - cy) / fy * rz

        plane = _solve_plane(rx.tolist(), ry.tolist(), rz.tolist())
        if plane is None:
            return None
        a, b, c = plane

        # One MAD-trim of ring residuals, then refit.
        resid = a * rx + b * ry + c - rz
        med = _median(resid.tolist())
        absdev = np.abs(resid - med)
        mad = _median(absdev.tolist())
        if mad > _MAD_EPS:
            keep = absdev <= p.mad_k * mad
            if int(keep.sum()) >= 3:
                kx, ky, kz = rx[keep], ry[keep], rz[keep]
                refit = _solve_plane(kx.tolist(), ky.tolist(), kz.tolist())
                if refit is not None:
                    a, b, c = refit
                    rx, ry, rz = kx, ky, kz

        resid_final = a * rx + b * ry + c - rz
        plane_residual = float(np.sqrt(np.mean(resid_final * resid_final))) if rz.size else 0.0

        # Coverage gate.
        food_valid = food & valid
        coverage = int(food_valid.sum()) / food_total
        if coverage < p.min_coverage:
            return None

        # Integrate V = (1/(fx·fy)) Σ max(0, z_plane − z) · z².
        fv, fu = np.nonzero(food_valid)
        fz = depth[fv, fu]
        fx_world = (fu.astype(np.float64) - cx) / fx * fz
        fy_world = (fv.astype(np.float64) - cy) / fy * fz
        z_plane = a * fx_world + b * fy_world + c
        h_raw = z_plane - fz
        # Plausibility: real food rises modestly above its plate. With no near support
        # plane (object over a far wall) the median height is huge → reject so V can't
        # explode into impossible mass/calories.
        heights = np.maximum(0.0, h_raw)
        if heights.size == 0 or float(np.median(heights)) > p.max_food_height_mm:
            return None
        h = np.where(h_raw > p.height_clamp_mm, h_raw, 0.0)
        sum_term = float(np.sum(h * fz * fz))
        volume_cm3 = (sum_term / (fx * fy)) / 1000.0
        return DepthVolumeResult(
            volume_cm3=volume_cm3, coverage=coverage, plane_residual_mm=plane_residual
        )

    def integrate_sample(self, sample, params: DepthVolumeParams | None = None) -> DepthVolumeResult | None:
        """Convenience over a :class:`DepthSample` (mirrors the Swift overload)."""
        return self.integrate(
            sample.depth_mm, sample.mask, sample.width, sample.height,
            sample.fx, sample.fy, sample.cx, sample.cy, params,
        )
